flamedisx/block_source.py

```diff
@@ -0,0 +1,341 @@
+
+            for x in (_this_block['model_functions']
+                      + _this_block['special_model_functions']
+                      + _this_block['static_attributes']):
+
+                # If a source attribute was specified,
+                # override the block's attribute.
+                if hasattr(self, x):
+                    setattr(b, x, getattr(self, x))
+
+                # Set the source attribute from the block's attribute
+                setattr(self, x, getattr(b, x))
+
+                # Someone might try to modify the source attribute after
+                # instantiation, then be really surprised when nothing happens.
+                # Sorry. I can't use properties to prevent writing, since
+                # those are class-bound.
+
+        # The source may declare additional frozen data methods
+        collected['frozen_data_methods'] += self.frozen_data_methods
+
+        # For array columns, the source may declare new columns
+        # or override the length of the old columns
+        for column, length in self.array_columns:
+            for i, (_old_column, _) in enumerate(collected['array_columns']):
+                if _old_column == column:
+                    # Change length of existing column
+                    collected['array_columns'][i] = (column, length)
+                    break
+            else:
+                # New column
+                collected['array_columns'] += [(column, length)]
+
+        # TODO: Ugly since source's conventions / naming is a bit divergent
+        self.data_methods = tuple(collected['model_functions']
+                                  + collected['special_model_functions'])
+        self.special_data_methods = tuple(collected['special_model_functions'])
+        self.frozen_data_methods = tuple(collected['frozen_data_methods'])
+        self.array_columns = tuple(collected['array_columns'])
+        # Only for inspection by user, not used in flamedisx
+        self.static_attributes = tuple(collected['static_attributes'])
+
+        self.inner_dimensions = tuple([
+            d for d in collected['dimensions']
+            if ((d not in self.final_dimensions)
+                and (d not in self.model_blocks[0].dimensions))])
+        self.initial_dimensions = self.model_blocks[0].dimensions
+
+    @staticmethod
+    def _find_block(blocks,
+                    has_dim: ty.Union[list, tuple, set],
+                    exclude: Block = None):
+        """Find a block with a dimension in allowed
+        Return ((dimensions, b), dimension found), or
+         raises BlockNotFoundError if no such block found.
+        """
+        for dims, b in blocks.items():
+            if b is exclude:
+                continue
+            for d in dims:
+                if d in has_dim:
+                    return (dims, b), d
+        raise BlockNotFoundError(f"No block with {has_dim} found!")
+
+    def _differential_rate(self, data_tensor, ptensor):
```

Comment:
This is an interesting thought; after factoring out the matrix multiply of block results into a separate function, the code is smaller and more readable now. But if you have a way of doing it more elegantly, or in a way that supports more general graphs, it would certainly be worth trying...

Reply:
Ok, thanks, we might do this in a future PR.